ddd/ddd-02-conn-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,167 @@
+ddd-02-conn-nonblocking
+=======================
+
+In addition to the changes in ddd-01-conn-blocking, the following issues arise:
+
+  - The application must be able to poll. This means that it must have some
+    OS-level synchronization object which is suitable for polling (e.g. a file
+    descriptor). The application may well want to poll other things than libssl
+    at the same time, so this needs to be something which can be polled
+    simultaneously with other things, not some special "poll" call provided
+    by libssl. The only thing meeting this criterion for varied POSIX systems is
+    a file descriptor.
+
+  - A valid design question is do we want to guarantee that the application
+    only needs to poll a single FD to determine readiness of a given
+    connection? It is possible we might have some reason to keep multiple
+    FDs internally (multiple UDP sockets etc.) but this means we need to have
+    a mechanism to tell the application some arbitrary number of FDs need to be
+    polled.
+
+    My rough view is that:
+
+    - we don't need more than one UDP socket TX-wise, because we can direct
+      destination addresses with sendto/sendmsg/sendmmsg.
+
+    - do we need more than one UDP socket RX-wise? If we bind to 0.0.0.0
+      there is no issue, and if an application wants to bind to a specific
+      interface it presumably does so for a reason. But what if an application
+      wants to bind to a specific interface (say for example, 192.0.2.1) and
+      later wants to do connection migration?
+
+      This seems like a relatively rare thing. Usually only servers would want
+      to bind to a specific interface and are less likely to do migration.
+
+      Some OSes provide means to aggregate FDs for purposes of polling. For
+      example, on Linux, I believe you can add FDs to an epoll FD and then call
+      poll(2) on that(?). Apparently only POLLIN is reported for this however so
+      this probably would not match expectations. Moreover we probably shouldn't
+      plan on this being available on all OSes.
+
+      PROPOSED SOLUTION: To minimise necessary changes, our best hope is
+      probably to plan to require the application to poll only a single FD. If
+      we ever end up needing multiple sockets, we can provide an advanced
+      interface which applications can choose to implement if they wish. In the
+      worst case they can forego some functionality if they don't implement it.
+      But to be clear, there is no immediate identified need for multiple
+      sockets at this time.
+
+  - The application needs to know when to listen for each of
+    POLLIN/POLLOUT/POLLERR on the file descriptor it has been told about.
+
+    CHANGE: The function `BIO_get_fd` is changed to `BIO_get_poll_fd` as the
```

Comment:
Not sure I understand the reasoning behind this. In what way is the fd returned by `BIO_get_poll_fd` different from the fd returned by `BIO_get_fd`?

Reply:
The idea is that we may not want to guarantee anything about the FD that should be polled (whether it is a socket, etc.) in future. If the API contract is just "this is a pollable FD, trying to do anything else with it is undefined", this keeps our  options open.

Reply:
Of course, there's a valid question as to whether this is a useful option to keep open. Are we likely, in future, to shift to some different QUIC design such that it's useful to return some unspecified non-socket FD here? (Think of the things the Linux kernel offers, eventfds, timerfds, etc.) Possibly not, though if we discover new requirements after MVP (for advanced features like connection migration etc.), it seems like a conceivable possibility. API design instincts tell me we should keep this as vague as possible if the client doesn't need a specific guarantee.

Reply:
Ok, this makes sense. It's quite similar to the ASYNC "wait_fd" mechanism that we have:

https://www.openssl.org/docs/manmaster/man3/ASYNC_WAIT_CTX_set_wait_fd.html